crates/log-server/src/rocksdb_logstore/writer.rs

```diff
@@ -178,13 +176,15 @@ impl LogStoreWriter {
         write_batch: &mut WriteBatch,
         buffer: &mut BytesMut,
     ) {
-        buffer.reserve(store_message.estimated_encode_size());
         let mut offset = store_message.first_offset;
         for payload in store_message.payloads.iter() {
             let key_bytes =
-                DataRecordKey::new(store_message.header.loglet_id, offset).encode_and_split(buffer);
-            let value_bytes = DataRecordEncoder::from(payload).encode_to_disk_format(buffer);
+                DataRecordKey::new(store_message.header.loglet_id, offset).to_binary_array();
+            let encoder = DataRecordEncoder::from(payload);
+            buffer.reserve(encoder.estimated_encode_size());
```

Comment:
This avoid unnecessary memory ballooning since it'll try to allocate a single buffer that's big enough for all records combined although we just need one big enough for the biggest record.